### PR TITLE
Use solidity delegatecall

### PR DESCRIPTION
**Pull request by d-xo**
Since version 0.5.0, the solidity `delegatecall` returns the result of the call, so the assembly implementation from `ds-proxy` is no longer needed.

@rainbreak you had some concerns about using the solidity `delegatecall` when we discussed this previously. What needs to be done in terms of bytecode auditing etc.?

### CODE REVIEW

File: src/pause.sol [hunk 1]
```diff
@@ -74,30 +74,18 @@ contract DSPause {
...
-        returns (bytes memory response)
+        returns (bytes memory)
     {
         bytes32 id = tag(guy, data, when);
 
         require(now >= when + delay, "ds-pause: delay not passed");
         require(scheduled[id] == true, "ds-pause: unscheduled execution");
 
         scheduled[id] = false;
+        (bool success, bytes memory response) = guy.delegatecall.gas(gasleft() - 1250)(data);
```
Comment (d-xo) [inline src/pause.sol]: The tests fail with `OutOfGas` errors if I don't have this `.gas(gasleft() - 1250)` statement.

I don't understand why? How is gas forwarded here? Does `delegatecall` consume all the gas that it is passed? I couldn't see anything in the docs describing how this works.... :disappointed:

Comment (asymmetric) [reply src/pause.sol]: I guess it has something to do with [this](https://github.com/ethereum/wiki/wiki/Subtleties):

> A CALL or CREATE can consume at most 63/64 of the gas remaining at the time the CALL is made; if a CALL asks for more than this prescribed maximum, then the inner call will only have the prescribed maximum gas regardless of how much gas was asked for.

`gasleft() / 64 * 63` works in fact.

It's not clear to me though why this doesn't just silently succeed, since it should provide the called contract the prescribed maximum anyway.

/cc @MrChico @livnev

Comment (gbalabasquer) [reply src/pause.sol]: As far as I understand the gas you rest from the `gasleft()` is the necessary to execute the `delegatecall` function itself (add, sub, mload operations). Otherwise you would be forwarding all the gas to execute the delegated code and nothing remains for doing the previous operations. In `ds-proxy` we are leaving 5K gas for this to have an important gap just in case in the future the gas costs increment.
There was some discussion about it in this PR: https://github.com/dapphub/ds-proxy/pull/18 )(I was actually confused about its purpose in that moment).

**Comment by asymmetric** [comment]
Maybe "delay not passed" could sound better with "delay not elapsed"? :)

**Review by asymmetric — COMMENTED** [review]
Maybe "delay not passed" could sound better with "delay not elapsed"? :)

File: src/pause.sol (hunk 1 above)
Comment (asymmetric) [reply src/pause.sol]: [This](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-150.md) is the actual EIP:

> If a call asks for more gas than the maximum allowed amount (i.e. the total amount of gas remaining in the parent after subtracting the gas cost of the call and memory expansion), do not return an OOG error; instead, if a call asks for more gas than all but one 64th of the maximum allowed amount, call with all but one 64th of the maximum allowed amount of gas (this is equivalent to a version of EIP-901 plus EIP-1142). CREATE only provides all but one 64th of the parent gas to the child call.

So it seems to me that hevm should not return an OOG error here.

I think this whole 63/64 thing is meant to address security concerns, and replaces the stack limit:

> replacing the call stack depth limit with a "softer" gas-based restriction, thereby eliminating call stack depth attacks as a class of attack that contract developers have to worry about and hence increasing contract programming safety.

**Comment by rainbreak** [comment]
This is pending the implementation of delegatecall in hevm being fixed. There is currently a mismatch between the yellow paper and clients / the jello paper, and hevm follows the yellow paper. @livnev

**Comment by gbalabasquer** [comment]
Btw I'd use the assembly version regardless the new solidity implementation. It is well tested and working and also we got to handle the error messaging correctly. I remember doing some quick test with the solidity one and couldn't get the error messaging when the called function reverted.

**Comment by rainbreak** [comment]
hevm issues should be fixed with https://github.com/dapphub/dapptools/pull/158

**Comment by d-xo** [comment]
I would like to either merge this change or pull the `delegatecall` implementation out into a mixin (`ds-calls` or smth similar). 

I agree with @nmushegian that copy pasting around a bunch of hard to parse assembly every time we want to do a `delegatecall` is suboptimal (see https://github.com/dapphub/ds-pause/issues/15).

@gbalabasquer I wrote a lil test [here](https://github.com/dapphub/ds-pause/commit/d00fbed4e3ca207c7aa152be2f8df1179d116c41) for the error message propagation and it seems to be working OK (results [here](https://travis-ci.com/dapphub/ds-pause/builds/109126794#L993)) to me, am I missing something?

**Comment by gbalabasquer** [comment]
@xwvvvvwx hmmm but I do not see the original `require` message being propagated as the `delegatecall` failure. Actually it is returning `ds-pause-delegatecall-error` as the final error message.
What we do in `ds-proxy` is to throw as error at the `proxy.execute` level the same error that the `delegatecall` source threw.

**Comment by gbalabasquer** [comment]
This is the test case: https://github.com/dapphub/ds-proxy/blob/master/src/proxy.t.sol#L241
Anyway with the addition of ProxyDelegator I'm not very sure if we need to overcomplicate this.

**Comment by d-xo** [comment]
@gbalabasquer got it now, thx for the clarification :pray: 

After thinking it over for a while, I'm going to merge this based on the following:

- The benefits to readability are substantial
- Usage of the assembly implementation does not mean we no longer rely on `solc`, but instead means we rely on the safety of the assembly -> bytecode compiler instead of the solidity -> bytecode compiler (admittedly this should be a much simpler piece of code).
- The new solidity `delegatecall` implementation has been live since November 2018 and been used in most contracts deployed since then (presumably) without issue.
- We can formally verify the behavior of the generated bytecode in the context of the contracts that we expect it to be used with (moms & the pause itself)

File: src/pause.sol [hunk 2]
```diff
@@ -96,24 +96,12 @@ contract DSPauseProxy {
...
-            returndatacopy(add(response, 0x20), 0, size)
-
-            switch iszero(succeeded)
-            case 1 {
-                revert(add(response, 0x20), size)
-            }
-        }
+        bool ok;
+        (ok, out) = usr.delegatecall(fax);
+        require(ok, "ds-pause-delegatecall-error");
```
Comment (rainbreak) [inline src/pause.sol:105]: N.b. This `require` behaviour is different. in the assembly version the error from the delegatecall was being propagated, but here you are returning a string.
